README.md
<div align='center'>

# Koa Isomorphic Router
---

[![Build Status][travis-img]][travis-url]
[![Coverage Status][coverage-img]][coverage-url]
[![NPM version][npm-badge]][npm-url]
![Code Size][code-size-badge]
[![License][license-badge]][license-url]
[![PR's Welcome][pr-welcoming-badge]][pr-welcoming-url]

</div>

<!-- ***************** -->

[travis-img]: https://travis-ci.org/3imed-jaberi/koa-isomorphic-router.svg?branch=master
[travis-url]: https://travis-ci.org/3imed-jaberi/koa-isomorphic-router
[coverage-img]: https://coveralls.io/repos/github/3imed-jaberi/koa-isomorphic-router/badge.svg?branch=master
[coverage-url]: https://coveralls.io/github/3imed-jaberi/koa-isomorphic-router?branch=master
[npm-badge]: https://img.shields.io/npm/v/koa-isomorphic-router.svg?style=flat
[npm-url]: https://www.npmjs.com/package/koa-isomorphic-router
[license-badge]: https://img.shields.io/badge/license-MIT-green.svg?style=flat
[license-url]: https://github.com/3imed-jaberi/koa-isomorphic-router/blob/master/LICENSE
[code-size-badge]: https://img.shields.io/github/languages/code-size/3imed-jaberi/koa-isomorphic-router
[pr-welcoming-badge]: https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat
[pr-welcoming-url]: https://github.com/koajs/koa/pull/new

[trek-router]: https://github.com/trekjs/router
[echo-router]: https://github.com/labstack/echo
[support-url]: https://github.com/koajs/3imed-jaberi/koa-isomorphic-router
[path-to-regexp]: https://github.com/pillarjs/path-to-regexp
[route-recognizer]: https://github.com/tildeio/route-recognizer
[route-trie]: https://github.com/zensh/route-trie
[routington]: https://github.com/pillarjs/routington

[trek-router-benchmarks-url]: https://github.com/trekjs/router#benchmarks
[fastify-benchmarks-url]: https://github.com/fastify/benchmarks#benchmarks
[405-501-warn]: https://github.com/3imed-jaberi/koa-isomorphic-router/blob/master/index.js#L204
[trek-router-405-501-warn]: https://github.com/trekjs/router/issues/23
[trek-router-reg-exp]: https://github.com/trekjs/router/issues/24

<!-- ***************** -->

### The fastest elegant modern amiable Koa.js Router ⚡.


## `Features`

* 🦄 Based on top of [Trek Router][trek-router] which inspired by [Echo][echo-router]'s Router.
* 🚀 Faster than other Koa.js router solutions.
* 💅🏻 Express-style routing (`app.get`, `app.post`, `app.put`, `app.delete`, etc.)
* 🔥 Blaze and lightweight router.
* ⚖️ Tiny Bundle: less than 2.5kB (gzip)
* 🪁 Named URL parameters.
* 🎯 Route middleware.
* 🥞 Support router layer middlewares.
* 📋 Responds to `OPTIONS` requests with allowed methods.
* ⛔️ Support for `405 Method Not Allowed`.
* ❌ Support for `501 Path Not Implemented`.
* 🧼 Support `trailing slash` and `fixed path` by automatic redirection.
* ✨ Asynchronous support (`async/await`).
* 🎉 TypeScript support.

### `Note`

Currently, this modules support `405 Method Not Allowed` 
and `501 Path Not Implemented` for 'static' routes only 
as you can see [here][405-501-warn].

As soon as possible when I get response [here][trek-router-405-501-warn]
will support 'param' and 'match-any' routes.

For support `RegExp` in route path also I waiting for 
any response [here][trek-router-reg-exp].


## `Benchmarks`

All Koa router solutions depend on [path-to-regexp][] when our solution 
relies on the [trek-router][] which has the best performance and not over 
the [path-to-regexp][] only also on others such as ([route-recognizer][], 
[route-trie][], [routington][] ...etc).

- [x] See trek-router [benchmarks][trek-router-benchmarks-url] <small>(`trek-router` better perf. than `path-to-regexp`)</small>. <br />
- [x] See fastify [benchmarks][fastify-benchmarks-url] <small>(`koa-isomorphic-router` better perf. than `koa-router`)</small>.


## `Installation`

```bash
# npm
$ npm install koa-isomorphic-router
# yarn
$ yarn add koa-isomorphic-router
```


## `Usage`

This is a practical example of how to use.

```javascript
const Koa = require('koa')
const Router = require('koa-isomorphic-router')

const app = new Koa()
const router = Router()

router
  .get('/product/:id', (ctx, next) => {
    ctx.body = { productId: ctx.params.id }
  })

app.use(router.routes());

app.listen(5050);
```


## `API`

### new Router(options?)

Create a new router.

| Param | Type  | Description |
| ---   | ---   | ---         |
| [options] | `Object` |  |
| [options.prefix] | `String` | prefix router paths |
| [options.throw] | `Boolean` | throw error instead of setting status and header |
| [options.notImplemented] | `function` | throw the returned value in place of the default NotImplemented error |
| [options.methodNotAllowed] | `function` | throw the returned value in place of the default MethodNotAllowed error |

### router.get|post|put|patch|delete|all(path, ...middlewares)

The http methods provide the routing functionality in `router`.

Method middleware and handlers follow usual Koa middleware behavior,
except they will only be called when the method and path match the request.

```js
// handle a GET / request.
router.get('/', (ctx) => { ctx.body = 'Hello World!' })
```

### router.prefix(prePath)

Route paths can be prefixed at the router level:

```js
// handle a GET /prePath/users request.
router
  .prefix('/prePath')
  .get('/users', (ctx) => { ctx.body = 'Hello World!' })
```

### router.route(path)

Lookup route with given path.

```js
// handle a GET /users request.
router
  .route('/users')
  .get((ctx) => { ctx.body = 'Hello World!' })
```

### router.use(...middlewares)

Use given middleware(s). Currently, use middleware(s) for all paths of router isntance.

### router.routes()

Returns router middleware which handle a route matching the request.


## `Support`

If you have any problem or suggestion please open an issue [here][support-url].


## `Call for Maintainers/Contributors`

This module is a attempt to craft an isomorphic fast Router, from/to Koa.js 
community. So, don't hesitate to offer your help ❤️.


## `Contributors`

| Name            | Website                         |
| --------------- | ------------------------------- |
| **Imed Jaberi** | <https://www.3imed-jaberi.com/> |


#### License
---

[MIT](LICENSE) &copy;	[Imed Jaberi](https://github.com/3imed-jaberi)


CHANGELOG.md
# CHANGE LOG

### v1.0.0 (27 Jan 2021)
  * use ES6 features over ES5 like class over function.
  * replace `mnemonist/lru-cache` with `hashlru`.
  * add prefix method.
  * add prefix as option.
  * add route method.
  * add use method.
  * add throw as option.
  * add custom method not allowed function as option.
  * add custom path not implemented function as option.
  * use Symbol(s) to make some props as private.
  * ignore default favicon request.
  * clean up the test.
  * add typescript support.
  * update README.md docs.

### v0.2.0 (19 Jan 2021)
  * improve the perf.
  * remove the `.on()` method support.
  * early handle the `option` request.
  * support this methods [get|put|post|patch|delete|all] only.
  * add cache support using the fastes LRU cache `mnemonist/lru-cache`.

### v0.1.1 (16 Jan 2021)
  * fix node version inside the package.json
  * add benchmarks section inside the README.md

### v0.1.0 (16 Jan 2021)
  * initial release

index.spec.js
/*!
 * koa-isomorphic-router test
 */

'use-strict'

const Koa = require('koa')
const request = require('supertest')
const assert = require('assert')
const FastRouter = require('.')
const METHODS = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE']

describe('koa-isomorphic-router', () => {
  describe('self', () => {
    it('should return a object', () => {
      assert.strictEqual(typeof new FastRouter(), 'object')
    })

    it('should not return private props/method', () => {
      assert.deepStrictEqual(
        Object.getOwnPropertyNames(FastRouter.prototype),
        [
          'constructor', 'get', 'post', 'put', 'patch',
          'delete', 'all', 'prefix', 'route', 'use', 'routes'
        ]
      )
    })
  })

  describe('http verbs/methods', () => {
    function createKoaApp (method, path) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('get method', (done) => {
      request(createKoaApp('get', '/test').listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get data/)
        .expect(200, done)
    })

    it('post method', (done) => {
      request(createKoaApp('post', '/test').listen())
        .post('/test')
        .expect('Content-Type', /json/)
        .expect(/post data/)
        .expect(200, done)
    })

    it('put method', (done) => {
      request(createKoaApp('put', '/test').listen())
        .put('/test')
        .expect('Content-Type', /json/)
        .expect(/put data/)
        .expect(200, done)
    })

    it('patch method', (done) => {
      request(createKoaApp('patch', '/test').listen())
        .patch('/test')
        .expect('Content-Type', /json/)
        .expect(/patch data/)
        .expect(200, done)
    })

    it('delete method', (done) => {
      request(createKoaApp('delete', '/test').listen())
        .delete('/test')
        .expect('Content-Type', /json/)
        .expect(/delete data/)
        .expect(200, done)
    })

    describe('all method', () => {
      const app = createKoaApp('all', '/test')

      METHODS.forEach((method) => {
        it(method.toLowerCase(), (done) => {
          request(app.listen())[method.toLowerCase()]('/test').expect(200, done())
        })
      })
    })
  })

  describe('favicon request', () => {
    function createKoaApp (method, path) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should ignore the favicon request', (done) => {
      request(createKoaApp('get', '/test').listen())
        .get('/favicon.ico')
        .expect(404, done)
    })
  })

  describe('allow header field', () => {
    function createKoaApp (method, path, methods) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      methods.forEach(m => { router[m](path, () => {}) })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should allow header funcs work', (done) => {
      request(createKoaApp(
        'get',
        'test',
        ['post', 'put', 'delete']
      ).listen()).get('/test').expect(200, done)
    })
  })

  describe('params', () => {
    function createKoaApp (method, path) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data ${ctx.params ? ctx.params.state : ''}` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should return a response with params', (done) => {
      request(createKoaApp('get', '/test/:state').listen())
        .get('/test/work')
        .expect(/get data work/)
        .expect(200, done)
    })
  })

  describe('trailing slash and fixed path', () => {
    function createKoaApp (method, path) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should normalize the path and redirect to correct one', (done) => {
      request(createKoaApp('get', '//test/').listen())
        .get('//test')
        .expect(301)
        .expect('Location', '/test')
        .end(done)
    })
  })

  describe('options request', () => {
    function createKoaApp (method, path) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should responds with allowed methods', (done) => {
      request(createKoaApp('get', '/test').listen())
        .options('/test')
        .expect('Allow', 'GET')
        .expect(204, done)
    })
  })

  describe('405 method not allowed', () => {
    function createKoaApp (method, path, throwObj) {
      // init
      const app = new Koa()
      const router = new FastRouter(throwObj)

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should responds with method not allowed', (done) => {
      request(createKoaApp('get', '/test').listen())
        .post('/test')
        .expect(405, done)
    })

    it('should throw when passed throw eq true as instance arg', (done) => {
      request(createKoaApp('get', '/test', { throw: true }).listen())
        .post('/test')
        .expect(405, (err) => err ? done(err) : done())
    })

    it('should throw with custom function when passed as instance arg', (done) => {
      function methodNotAllowed () {
        const methodNotAllowedErr = new Error('Custom Method Not Allowed Error')
        methodNotAllowedErr.type = 'custom'
        methodNotAllowedErr.statusCode = 405
        methodNotAllowedErr.body = {
          error: 'Custom Method Not Allowed Error',
          statusCode: 405,
          otherStuff: true
        }

        return methodNotAllowedErr
      }

      request(createKoaApp('get', '/test', { throw: true, methodNotAllowed }).listen())
        .post('/test')
        .expect(405)
        .end((err) => err ? done(err) : done())
    })
  })

  describe('501 path not implemented', () => {
    function createKoaApp (method, path, throwObj) {
      // init
      const app = new Koa()
      const router = new FastRouter(throwObj)

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should responds with path not implemented', (done) => {
      request(createKoaApp('get', '/test').listen())
        .get('/not-imp-test')
        .expect(501, done)
    })

    it('should throw when passed throw eq true as instance arg', (done) => {
      request(createKoaApp('get', '/test', { throw: true }).listen())
        .get('/not-imp-test')
        .expect(501)
        .end((err) => err ? done(err) : done())
    })

    it('should throw with custom function when passed as instance arg', (done) => {
      function notImplemented () {
        const pathNotImplementedErr = new Error('Custom Path Not Implemented Error')
        pathNotImplementedErr.type = 'custom'
        pathNotImplementedErr.statusCode = 501
        pathNotImplementedErr.body = {
          error: 'Custom Path Not Implemented Error',
          statusCode: 501,
          otherStuff: true
        }

        return pathNotImplementedErr
      }

      request(createKoaApp('get', '/test', { throw: true, notImplemented }).listen())
        .get('/not-imp-test')
        .expect(501)
        .end((err) => err ? done(err) : done())
    })
  })

  describe('cache', () => {
    function createKoaApp (method, path) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router[method](path, (ctx) => {
        ctx.status = 200
        ctx.body = { msg: `${method} data` }
      })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should get request from the cache', (done) => {
      const app = createKoaApp('get', '/test')

      // no cached request.
      request(app.listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, () => {
          // cached request.
          request(app.listen())
            .get('/test')
            .expect('Content-Type', /json/)
            .expect(/get/)
            .expect(200, done)
        })
    })
  })

  describe('route method', () => {
    function createKoaApp (method, path) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router
        .route(path)[method]((ctx) => {
          ctx.status = 200
          ctx.body = { msg: `${method} data` }
        })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should route method work', (done) => {
      request(createKoaApp('get', '/test').listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })
  })

  describe('prefix as instance arg and method', () => {
    function createKoaApp (method, path, { prefix, useMethod }) {
      // init
      const app = new Koa()
      const router = new FastRouter(!useMethod ? { prefix } : undefined)

      // register a route
      router
        .prefix(useMethod ? prefix : undefined)[method](path, (ctx) => {
          ctx.status = 200
          ctx.body = { msg: `${method} data` }
        })

      // add router middelware
      app.use(router.routes())

      return app
    }

    // bad arg passed ==> don't use the prefix.

    it('should prefix instance arg work', (done) => {
      request(createKoaApp(
        'get',
        '/test',
        { prefix: '/preRoute', useMethod: false }
      ).listen())
        .get('/preRoute/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })

    it('should prefix method work', (done) => {
      request(createKoaApp(
        'get',
        '/test',
        { prefix: '/preRoute', useMethod: true }
      ).listen())
        .get('/preRoute/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })
  })

  describe('use method', () => {
    function createKoaApp (method, path, handler) {
      // init
      const app = new Koa()
      const router = new FastRouter()

      // register a route
      router
        .use(handler)[method](path, (ctx) => {
          ctx.status = 200
          ctx.body = { msg: `${method} data` }
        })

      // add router middelware
      app.use(router.routes())

      return app
    }

    it('should use method work with good arg', (done) => {
      const handler = (ctx, next) => {
        console.log('logger', ctx.status, ctx.path)
        return next
      }

      request(createKoaApp('get', 'test', handler).listen())
        .get('/test')
        .expect('Content-Type', /json/)
        .expect(/get/)
        .expect(200, done)
    })

    it('should use method throw with bad arg', () => {
      assert.throws(() => { createKoaApp('get', 'test', 'bad args') })
    })
  })
})


index.d.ts
/* ====================== USAGE ======================
    import * as Router from "koa-isomorphic-router";
    const router = new Router();
 ===================================================== */

import * as Koa from "koa";

declare namespace Router {
    export interface IRouterOptions {
        /**
         * prefix for all routes.
         */
        prefix?: string;
        /**
         * throw error instead of setting status and header
         */
        throw?: boolean;
        /**
         * throw the returned value in place of the default NotImplemented error
         */
        notImplemented?: () => any;
        /**
         * throw the returned value in place of the default MethodNotAllowed error
         */
        methodNotAllowed?: () => any;
    }

    export type IMiddleware<StateT = any, CustomT = {}> = Koa.Middleware<StateT, CustomT>
}
 
declare class Router<StateT = any, CustomT = {}> {
    /**
     * Create a new router.
     */
    constructor(options?: Router.IRouterOptions);

    /**
     * HTTP get method
     */
    get(
        path?: string,
        ...middlewares: Array<Router.IMiddleware<StateT, CustomT>>
    ): Router<StateT, CustomT>;

    /**
     * HTTP post method
     */
    post(
        path?: string,
        ...middlewares: Array<Router.IMiddleware<StateT, CustomT>>
    ): Router<StateT, CustomT>;

    /**
     * HTTP put method
     */
    put(
        path?: string,
        ...middlewares: Array<Router.IMiddleware<StateT, CustomT>>
    ): Router<StateT, CustomT>;

    /**
     * HTTP patch method
     */
    patch(
        path?: string,
        ...middlewares: Array<Router.IMiddleware<StateT, CustomT>>
    ): Router<StateT, CustomT>;

    /**
     * HTTP delete method
     */
    delete(
        path?: string,
        ...middlewares: Array<Router.IMiddleware<StateT, CustomT>>
    ): Router<StateT, CustomT>;

    /**
     * Register route with all methods.
     */
    all(
        path?: string,
        ...middlewares: Array<Router.IMiddleware<StateT, CustomT>>
    ): Router<StateT, CustomT>;

    /**
     * Set the path prefix for a Router instance that was already initialized.
     */
    prefix(prefix: string): Router<StateT, CustomT>;

    /**
     * Lookup route with given `path`.
     */
    route(path: string): Router<StateT, CustomT>;

    /**
     * Use given middleware.
     */
    use(...middleware: Array<Router.IMiddleware<StateT, CustomT>>): Router<StateT, CustomT>;

    /**
     * Returns router middleware which dispatches a route matching the request.
     */
    routes(): Router.IMiddleware<StateT, CustomT>;
}

export = Router;


index.js
/*!
 * koa-isomorphic-router
 *
 *
 * Copyright(c) 2021 Imed Jaberi
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 */

const FastRouter = require('trek-router')
const koaCompose = require('koa-compose')
const hashlruCache = require('hashlru')

/**
 * Symbol(s) used to set some attributes and methods as private.
 *
 * @api private
 */

// attributes.
const fastRouter = Symbol('@@fastRouter$$')
const METHODS = Symbol('@@METHODS$$')
const throws = Symbol('@@throws$$')
const methodNotAllowed = Symbol('@@methodNotAllowed$$')
const notImplemented = Symbol('@@notImplemented$$')
const routePrefix = Symbol('@@routePrefix$$')
const routePath = Symbol('@@routePath$$')
const middlewaresStore = Symbol('@@middlewaresStore$$')
const cache = Symbol('@@cache$$')
const allowHeaderStore = Symbol('@@allowHeaderStore$$')

// methods.
const on = Symbol('@@on$$')
const getAllowHeaderTuple = Symbol('@@getAllowHeaderTuple$$')
const normalizePath = Symbol('@@normalizePath$$')

/**
 * Fast and isomorphic Router for Koa.js.
 *
 * @api public
 */

class Router {
  // init Router.
  constructor (options = {}) {
    // init attributes.
    this[fastRouter] = new FastRouter()
    this[METHODS] = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE']
    this[throws] = typeof options.throw === 'boolean' ? options.throw : false
    this[methodNotAllowed] = options.methodNotAllowed
    this[notImplemented] = options.notImplemented
    this[routePrefix] = typeof options.prefix === 'string' ? options.prefix : '/'
    this[routePath] = undefined
    this[middlewaresStore] = []
    this[cache] = hashlruCache(1000)
    this[allowHeaderStore] = [] // [{ path: '', methods: [] }]
  }

  // normalize the path by remove all trailing slash.
  [normalizePath] (path) {
    path = path.replace(/\/\/+/g, '/')
    if (path !== '/' && path.slice(-1) === '/') {
      path = path.slice(0, -1)
    }

    return path
  }

  // get allow header for specific path.
  [getAllowHeaderTuple] (path) {
    return this[allowHeaderStore].find(allow => allow.path === path)
  }

  // register route with specific method.
  [on] (method, path, ...middlewares) {
    // handle the path arg when passed as middleware.
    if (typeof path !== 'string') {
      middlewares = [path, ...middlewares]
      path = this[routePath]
    }

    // normalize the path.
    path = this[normalizePath](this[routePrefix] + path)

    // register path with method(s) to re-use as allow header filed.
    // allow header.
    const allow = this[getAllowHeaderTuple](path)

    // stock to allow header store with unique val array.
    this[allowHeaderStore] = [...new Map([
      ...this[allowHeaderStore],
      { path, methods: !allow ? [method] : [...new Set([...allow.methods, method])] }
    ].map(item => [item.path, item])).values()]

    // register to route to the trek-router stack.
    this[fastRouter].add(method, path, koaCompose(middlewares))

    // give access to other method after use the current one.
    return this
  }

  // register route with get method.
  get (path, ...middlewares) {
    return this[on]('GET', path, ...middlewares)
  }

  // register route with post method.
  post (path, ...middlewares) {
    return this[on]('POST', path, ...middlewares)
  }

  // register route with put method.
  put (path, ...middlewares) {
    return this[on]('PUT', path, ...middlewares)
  }

  // register route with patch method.
  patch (path, ...middlewares) {
    return this[on]('PATCH', path, ...middlewares)
  }

  // register route with delete method.
  delete (path, ...middlewares) {
    return this[on]('DELETE', path, ...middlewares)
  }

  // `router.all()` method >> register route with all methods.
  all (path, ...middlewares) {
    return this[on](this[METHODS], path, ...middlewares)
  }

  // add prefix to route path.
  prefix (prefix) {
    this[routePrefix] = typeof prefix === 'string' ? prefix : this[routePrefix]
    return this
  }

  // give access to write once the path of route.
  route (path) {
    // update the route-path.
    this[routePath] = path

    // give access to other method after use the current one.
    return this
  }

  // use given middleware, if and only if, a route is matched.
  use (...middlewares) {
    // check middlewares.
    if (middlewares.some(mw => typeof mw !== 'function')) {
      throw new TypeError('".use()" requires a middleware(s) function(s)')
    }

    // add the current middlewares to the store.
    this[middlewaresStore] = [...this[middlewaresStore], ...middlewares]

    // give access to other method after use the current one.
    return this
  }

  // router middleware which handle a route matching the request.
  routes () {
    return async (ctx) => {
      // normalize the path.
      const path = this[normalizePath](ctx.path)

      // ignore favicon request.
      // src: https://github.com/3imed-jaberi/koa-no-favicon/blob/master/index.js
      if (/\/favicon\.?(jpe?g|png|ico|gif)?$/i.test(ctx.path)) { return }

      // init route matched var.
      let route

      // have slashs ~ solve trailing slash.
      if (path !== ctx.path) {
        ctx.response.status = 301
        ctx.redirect(`${path}${ctx.search}`)
        return
      }

      // generate the cache key.
      const requestCacheKey = `${ctx.method}_${ctx.path}`
      // get the route from the cache.
      route = this[cache].get(requestCacheKey)

      // if the current request not cached.
      if (!route) {
        // find route inside the routes stack.
        route = this[fastRouter].find(ctx.method, ctx.path)
        // put the matched route inside the cache.
        this[cache].set(requestCacheKey, route)
      }

      // extract the handler func and the params array.
      const [handler, routeParams] = route

      // check the handler func isn't defined.
      if (!handler) {
        // warning: need more work with trek-router for support 'param' and 'match-any' route.

        // get methods exist on allow header.
        const allowHeaderFiled = this[getAllowHeaderTuple](path)

        if (allowHeaderFiled) {
          // if `OPTIONS` request responds with allowed methods.
          if (ctx.method === 'OPTIONS') {
            ctx.status = 204
            ctx.set('Allow', allowHeaderFiled.methods.join(', '))
            ctx.body = ''
            return
          }

          // support 405 method not allowed.
          if (this[throws]) {
            throw typeof this[methodNotAllowed] === 'function'
              ? this[methodNotAllowed]()
              : (() => {
                const notAllowedError = new Error(`"${ctx.method}" is not allowed in "${ctx.path}".`)
                notAllowedError.statusCode = 405

                return notAllowedError
              })()
          }

          ctx.status = 405
          ctx.set('Allow', allowHeaderFiled.methods.join(', '))
          ctx.body = `"${ctx.method}" is not allowed in "${ctx.path}".`
          return
        }

        // suport 501 path not implemented.
        if (this[throws]) {
          throw typeof this[notImplemented] === 'function'
            ? this[notImplemented]()
            : (() => {
              const notImplError = new Error(`"${ctx.path}" not implemented.`)
              notImplError.statusCode = 501

              return notImplError
            })()
        }

        ctx.status = 501
        ctx.set('Allow', '')
        ctx.body = `"${ctx.path}" not implemented.`
        return
      }

      // check if the route params isn't empty array.
      if (routeParams.length > 0) {
        // parse the params if exist.
        const params = {}
        routeParams.forEach(({ name: key, value }) => { params[key] = value })

        // append params to ctx and ctx.request.
        ctx.params = ctx.request.params = params
      }

      // wait to all middlewares stored by the `use` method.
      await Promise.all(this[middlewaresStore])

      // wait the handler.
      await handler(ctx)
    }
  }
}

/**
 * Expose `Router()`.
 */

module.exports = Router
